src/goto-instrument/contracts/contracts.cpp

```diff
@@ -221,8 +202,41 @@ void code_contractst::check_apply_loop_contracts(
     return invariant_copy;
   };
```

Comment:
The closure if for processing quantifiers correctly. Every time we use the invariant expression, we have to use a “fresh” expression because otherwise the temporary variables introduced for quantifiers will clash.

Consider `__CPROVER_loop_invariant ( __CPROVER_forall { int i ; … })`. When we process it, we don’t actually have a special scope for `i` in the goto program. We introduce a new temporary in the current scope and rewrite (replace `i`) in the invariant. If we don’t create a “fresh” `i` for each occurrence of the invariant expression (2 assertions, 1 assumption) then, for example, the temporary for `i` from base case assertion will clash with the one in the step case assertion

Reply:
No, you don't have to worry about the uniqueness of `i` within quantifiers: the back-end takes care of this, i.e., the scope of an identifier introduced within a quantifier is just that quantified expression. This was fixed in d9b6f156f69e.

Reply:
No, I believe this is different. I tried disabling this and using `invariant` expression directly everywhere and it fails exactly as it did before. This is my understanding:

For quantifiers that appear within `assert` and `assume` statements, the goto-cc compiler introduces new temporary variables during compilation. However, for quantifiers that appear within function and loop contracts, since the are never compiled, not until instrumentation, temporaries are not introduced during compilation. At the instrumentation step, we introduce a unique temporary for each unique quantified variable, in each quantified expression. I could just make 3 copies of the invariant, each with its own quantified variables, but I used this lamda to reduce code duplication and construct them on demand.

Reply:
@tautschnig Thanks! Yes indeed this works as expected now. Just introducing a single temporary for each unique quantified variable works fine now, even when thise quantified expression is used more than once.